### ISSUE TITLE
Site selection for Apps

### ISSUE THREAD

**Issue by djf604** [Parsl/parsl#215]
When the `sites=` keyword of the `@App` decorator is given a list with multiple values, the site assignment for that App is randomly chose from among the list. It would be good to respect the order given in the list.

For example, give the following code:

```python
@App('python', dfk, sites=['one', 'two'])
def _app():
    pass
```

It would be good if Parsl first tried to submit the app to site `one`, and if that site is full or fails for some reason, then try site `two`.

**Comment by yadudoc** [comment]
We currently have random as the only site selector and I can see the need for being able to select the site selector. For instance, as @djf604 mentioned we might want to say, "site_selector"="overflow", which makes a lot of sense for cases where you want to burst to the cloud when all "free" resources are unavailable. 

In swift we had the notion of a dynamic-score for sites, which determined the likelihood of jobs being assigned. In this case we might want to specify "site_selector"="site_priority", or even "dynamic_site_priority" where we update the priority scores based on information gathered during the run.

**Comment by annawoodard** [comment]
Agreed on adding more selection modes, I just wanted to make sure we don't lose the `shuffle` capability, because it has use cases.

**Comment by benclifford** [comment]
In the distant past with Swift, lots of different people wanted lots of different site selection modes - it would be neat if when going over a certain complexity level, this ended up as being something describable in Python, rather than as a bewildering configuration language.

**Comment by yadudoc** [comment]
Wanted to leave in a note that it might be interesting to make the site selection login pluggable. Even more fun to integrate something like @rohankumar42 's delta into this.
